### ISSUE TITLE
Cleanup needed.

### ISSUE THREAD

**Issue by cboulay** [sccn/xdf#2]
Now that the project is on github, it is not necessary to maintain multiple versions of the same files in different directories. The newer versions can be deleted, the repository committed, then [make a release](https://help.github.com/articles/creating-releases/). Then the newer versions can be restored, older deleted, commit, make a release.

With the releases made, a couple more things can be done to cleanup:
    \* The EEGLab import folders are entirely duplicates of the load_xdf folders with the exception of a few EEGLab specific files. This can probably be deduplicated.
    \* The binaries can be deleted. This is probably only important if the binaries will be updated often because git repositories bloat quickly when tracking changing binaries.

**Comment by cboulay** [comment]
@chkothe @dmedine 

I'm doing a dry run of the xdf tag & release on my own fork. [Please take a look](https://github.com/cboulay/xdf/releases).

I have a couple questions.

How is the structure?
I'm guessing that naming the folder `Matlab` is not going to make it easy to import this as a EEGLAB plugin. I haven't used EEGLAB plugins in a little while. Can you please remind me what a good folder name would be (preferably without version numbering)? Should I put this newly named folder as a child of the `Matlab` folder? I'm thinking of having `Python` as a sibling to the `Matlab` folder, hence the desire to keep a `Matlab` folder.

What do you think of the way the binaries are handled?

I pulled out the binaries from the repository and put them in the release as separately downloadable files. While this is overkill for binaries of this size (10 KB), it's a nice way to keep binaries out of the tracked repository while still making everything available in one place. Another good feature is that the binaries are available via static URLs so they can be coded into scripts.

The main problem with this approach is that it adds an extra step somewhere. Either the user has to download the binaries manually or we have to script it in somewhere. For example, we can have `load_xdf` look to see if the mex file exists and if it doesn't then it can do a 1-time download of the platform-appropriate mex file. If that fails (e.g., no network, wrong platform) then it can give instructions on how to compile. If we do that then I'll have to go back and edit `load_xdf.m` before making the tags and releases.

This approach might also work with LSL, only the binaries will be downloaded by the build system (e.g., cmake) instead of Matlab.

**Comment by chkothe** [comment]
Hi Chad,

thanks for cleaning that up! I agree, putting the binaries into releases is
the right thing to do.

So the EEGLAB import plugin folders are by convention named like:
xdfimportX.YY. It would be great if each release included an .zip file that
extracts to a folder with that name (so that users can drop that right into
their EEGLAB). So then, if we keep the version numbers out of the repo
paths, we'd need to have a step (either manual or scripted) to generate
that path prior to making a release. I think that's fine as long as it's
documented clearly. I would also really like to retain a separate .zip file
that contains only the raw MATLAB importer in pristine form without the
EEGLAB-related clutter (at some point we'll probably have similar raw xdf
importers in several more languages). In the repo itself, it would be great
if we could keep the raw MATLAB related files separated from the EEGLAB
import files (i.e., in their own directory) for the same reason, and one
way to do that (a bit unconventional but simple enough) would be to nest
the raw importer in a directory inside the EEGLAB importer, and add a hint
in the readme where to find the raw import source code. That could look
like this:

/MATLAB
  xdf
      load_xdf.m
      readme-loadxdf.txt
  eegplugin_xdfimport.m
  eeg_load_xdf.m
  pop_load_xdf.m
  readme-eegplugin.txt
/Python
  ...
creating-a-release.txt

Then the process for making a release could be to 1) copy the MATLAB folder
and rename to xdfimportX.YY, zip it up into an EEGLAB plugin release, and
to 2) zip up the content of the xdf folder into a raw MATLAB release, and
attach both files to the release in GitHub.

Keeping the binaries separate in the release package and downloading them
on the fly in load_xdf.m is a nice idea. This way, neither checking out the
repo nor creating a new release would require any manual step related to
those binaries.

Sounds agreeable?

Christian

**Comment by cboulay** [comment]
Instead of `creating-a-release.txt`, I may just be able to do `create_release.py` and use GitHub's [release API](https://developer.github.com/v3/repos/releases/#create-a-release).

Something like (pseudocode):

```
read load_xdf.m and parse version number
if git tag of version number exists:
    fail with message to increment version number

zip contents of Matlab\* into xdfimport<version#>.zip
zip contents of Matlab\* into xdf.zip

create release with tag vX.YY using API
Upload zip files as release assets.
Upload all mex files as release assets.
```

I'm a little too tired to do that tonight. I'll try again tomorrow.

**Comment by chkothe** [comment]
Well, that would be even better! :)

**Comment by cboulay** [comment]
Please take a look at the [release page](https://github.com/cboulay/xdf/releases) now. If everything looks OK then I will change the URLs from pointing to cboulay to pointing to sccn then redo the process from within a clone of the sccn repo.

Where would you like the instructions for make_release.py to go? A new .txt file? At the top of make_release.py?

```
github3 package is needed. Install with `pip install github3.py`
Change the version number in Matlab/xdf/load_xdf.m before making the release.
If a release with the same version number already exists on the server then the release will fail.
Push the latest commit to the server before making the release.
python make_release.py -u <ghusername> -p <ghpassword>
```

**Comment by dmedine** [comment]
This looks great! I don't have any opinion as to where the instructions 
go. It's up to you. Either an INSTRUCTIONS.txt or at the top of the .py 
makes sense to me.

We should also add xdfimport1.12 which is the version that EEGlab 
actually grabs to when you import xdfimport from that GUI. Also, I will 
talk to Ramon about changing the code in EEGLab to point here rather 
than to wherever xdfimport1.12 is (on our ftp somewhere, presumably).
-David

**Comment by chkothe** [comment]
Yes, that looks good! I'd say let's put the instructions at the top of the
.py file, since no one in their right might would (hopefully) run it
without looking at the source and/or help. By the way, it looks like some
.DS_Store files made it into the .zip files.

For the EEGLAB plugin we also need to make sure that eeg_load_xdf.m adds
the xdf sub-directory to the MATLAB path, as in (line 31):

% first load the .xdf file
addpath([fileparts(mfilename('fullpath')) filesep 'xdf']);
streams = load_xdf(filename);

**Comment by cboulay** [comment]
Yes, I had that in there then lost it through my tests. I'm sorry the mistake made it into the sccn commit. I'll fix it but I have to do a force push. Please don't clone between now and then.

**Comment by cboulay** [comment]
Or, I won't do a force push, I'll just delete the tags/releases, undo the 1.11 changes back to 1.08, modify, commit, then make a release again. It adds an extra couple commits but that's better than force push.

**Comment by cboulay** [comment]
I had the `addpath` in 1.08, but it got lost in 1.11. So I only need to fix 1.11.

I'm just going to delete tag/release for 1.11, then make the .m file change, commit, push, and make_release from a Windows PC.

I can't figure out where the .DS_Store files are coming from. They aren't in the directory (ls -all) before I run make_release.py but they show up in the zip. On top of that, make_release.py has a line in it to ignore .DS_Store if encountered and not add to zip. The only thing I can think of is that .DS_Store is being added to the zip by the OS during its creation.

Is it OK if the 1.08 zip files have .DS_Store files in them? Getting rid of those at this stage is probably more trouble than its worth.

**Comment by chkothe** [comment]
Ok, sounds good! Yes, let's gloss over the .DS_Store files for now, not a
big deal.

**Comment by cboulay** [comment]
The upload through make_release didn't work from my Windows computer. All the files were corrupted. It could be because that desktop has spotty wi-fi (but it's not that bad). Can someone test on their Windows computer? After looking over #4 , you can merge that then do your own `python make_release.py -u <username> -p <password>`

**Comment by cboulay** [comment]
Resolved when #4 was merged.
